### PR TITLE
#161504670 Implement login-functionality

### PR DESCRIPTION
**Pull request by augustineezinwa**
#### What does this PR do?
- This PR implements the login functionality.
#### Description of Task to be completed?
- add tests
- add controllers to login user
- add validations
- make error messsages more descriptive
#### How should this be manually tested?
- `npm test`
#### Any background context you want to provide?
- None
#### What are the relevant pivotal tracker stories?
-[ #161504670](https://www.pivotaltracker.com/story/show/161504670)
#### Screenshots (if appropriate)
screen shot 2018-11-02 at 7 19 41 am
screen shot 2018-11-02 at 7 20 24 am
screen shot 2018-11-02 at 7 21 10 am

#### Questions:
- N/A

[Finishes #161504670]

### CODE REVIEW

File: server/controllers/UserController.js [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+              message: ['Invalid email or password']
+            }
+          });
+        }
+        if (bcrypt.compareSync(password, userFound.password)) {
+          return res.status(200).json({
+            status: 'success',
+            message: 'you are logged in',
+            user: {
+              id: userFound.id,
```
Comment (KvNGCzA) [inline server/controllers/UserController.js]: I think you should destructure this object instead of using dot notation.

**Comment by sulenchy** [comment]
Awesome. The codebase is clean

File: server/controllers/UsersController.js [hunk 2]
```diff
@@ -45,6 +45,52 @@ class UsersController {
...
+    return User.findOne({
+      where: {
+        email
+      }
+    })
+      .then((userFound) => {
+        if (!userFound || !bcrypt.compareSync(password, userFound.password)) {
+          return res.status(401).json({
+            errors: {
+              message: ['Invalid email or password']
```
Comment (mbilesanmi) [inline server/controllers/UsersController.js]: Why are you using arrays for most of the error messages?

Comment (augustineezinwa) [reply server/controllers/UsersController.js]: The team agreed to follow as stated on the repository readme to display our errors. 
See link [https://github.com/andela/valinor-ah-backend#errors-and-status-codes](url)

Comment (mbilesanmi) [reply server/controllers/UsersController.js]: Let's discuss this during the TTL interaction

File: routes/api/users.js [hunk 3]
```diff
@@ -2,8 +2,12 @@ import express from 'express';
...
 import UserController from '../../server/controllers/UsersController';
 
-const { validateUserSignUp, checkExistingEmail } = UserValidation;
-
+const {
+  validateUserSignUp,
+  checkExistingEmail,
+  validateUserLogin
+} = UserValidation;
+const { userLogin } = UserController;
```
Comment (mbilesanmi) [inline routes/api/users.js]: For consistency, if you are destructuring this then do the same for signup on line 22.

File: server/config/config.js [hunk 4]
```diff
@@ -4,7 +4,7 @@ dotenv.config();
 
 export default {
   development: {
-    user: process.env.DB_USERNAME,
+    username: process.env.DB_USERNAME,
```
Comment (mbilesanmi) [inline server/config/config.js]: Why are we making this change?

File: test/server/controllers/UserController.test.js [hunk 5]
```diff
@@ -50,3 +53,109 @@ describe('/users/signup', () => {
       });
   });
 });
+
+describe('Testing Login feature -Integration testng', () => {
```
Comment (mbilesanmi) [inline test/server/controllers/UserController.test.js]:
```suggestion
describe('Testing Login feature -Integration testing', () => {
```

File: test/server/controllers/UserController.test.js [hunk 6]
```diff
@@ -50,3 +53,109 @@ describe('/users/signup', () => {
...
+  it(
+    'should display error message if user logins with empty field',
```
Comment (mbilesanmi) [inline test/server/controllers/UserController.test.js]:
```suggestion
    'should display error message if user logs in with empty field',
```

File: test/server/controllers/UserController.test.js [hunk 7]
```diff
@@ -50,3 +53,109 @@ describe('/users/signup', () => {
...
+      chai.request(app)
+        .post(loginUrl)
+        .send({
+          email: '',
+          password: ''
+        })
+        .end((err, res) => {
+          res.status.should.be.eql(422);
+          res.body.should.be.eql({
+            errors: {
```
Comment (mbilesanmi) [inline test/server/controllers/UserController.test.js]: Should this error return only for email or include password as well?

File: test/server/controllers/UserController.test.js [hunk 8]
```diff
@@ -50,3 +53,109 @@ describe('/users/signup', () => {
...
+              password: ['please enter password']
+            }
+          });
+        });
+      done();
+    }
+  );
+
+  it(
```
Comment (mbilesanmi) [inline test/server/controllers/UserController.test.js]: Whats the difference between this test and the test on line 141?

File: routes/api/users.js (hunk 3 above)
Comment (augustineezinwa) [reply routes/api/users.js]: Should I make this change on this pull request? it was introduced during the signup feature implementation.

File: test/server/controllers/UserController.test.js (hunk 7 above)
Comment (augustineezinwa) [reply test/server/controllers/UserController.test.js]: Ok. I will fix that.

File: test/server/controllers/UserController.test.js (hunk 8 above)
Comment (augustineezinwa) [reply test/server/controllers/UserController.test.js]: The test on line 80 is the scenario when the user logs in without providing a password. While the test on line 141 is the scenario where the user provides incorrect login credentials, be it password or email or both.

File: routes/api/users.js (hunk 3 above)
Comment (mbilesanmi) [reply routes/api/users.js]: Please, do that here.

File: server/config/config.js (hunk 4 above)
Comment (mbilesanmi) [reply server/config/config.js]: okay. noted

File: test/server/controllers/UserController.test.js (hunk 8 above)
Comment (mbilesanmi) [reply test/server/controllers/UserController.test.js]: `should display error message if user logs in with invalid credentials` and `should return error if user logs in with wrong password` are the tests I'm referring to.

Comment (augustineezinwa) [reply test/server/controllers/UserController.test.js]: Ok.

**Pull request merged**